pkg/azuredx/azureauth/auth.go

```diff
@@ -0,0 +1,196 @@
+		return r.Token, r.ExpiresOn, err
+	})
+	if err != nil {
+		return "", fmt.Errorf("service principal token unavailable: %w", err)
+	}
+	return "Bearer " + token, nil
+}
+
+// QueryDataAuthorization returns an HTTP Authorization-header value which
+// represents the respective request.
+func (c *ServiceCredentials) QueryDataAuthorization(ctx context.Context, req *backend.QueryDataRequest) (string, error) {
+	if c.QueryTimeout != 0 {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, c.QueryTimeout)
+		defer cancel()
+	}
+
+	switch {
+	case c.OnBehalfOf:
+		return c.queryDataOnBehalfOf(ctx, req)
+
+	default:
+		backend.Logger.Debug("using service principal token for data request")
+		return c.ServicePrincipalAuthorization(ctx)
+	}
+}
+
+func (c *ServiceCredentials) queryDataOnBehalfOf(ctx context.Context, req *backend.QueryDataRequest) (string, error) {
+	user := req.PluginContext.User // do nil-check once for all
+	if user == nil {
+		return "", errors.New("non-user requests not permitted with on-behalf-of configuration")
+	}
+
+	// Azure requires an ID token (instead of an access token) for the exchange.
+	userToken, ok := req.Headers["X-ID-Token"]
+	if !ok {
+		if _, ok := req.Headers["Authorization"]; !ok {
+			return "", errors.New("system accounts are denied with on-behalf-of configuration")
+		}
+
+		backend.Logger.Error("ID token absent for data request; enable them on the Azure portal under: “App Registrations” → the respective application → “Manage” → “Authentication”", "user", user.Login)
+		return "", errors.New("ID token absent for data request")
+	}
+
+	onBehalfOfToken, err := c.tokenCache.getOrSet(ctx, userToken, c.onBehalfOf)
+	if err != nil {
+		backend.Logger.Error(err.Error(), "user", user.Login)
+		// Don't leak any context to the end-user.
+		return "", errors.New("on-behalf-of token exchange failed for data request")
+	}
+	backend.Logger.Debug("aquired on-behalf-of token for data request")
+
+	return "Bearer " + onBehalfOfToken, nil
+}
+
+// OnBehalfOf resolves a token which impersonates the subject of userToken.
+// UserToken has to be an ID token. See the following link for more detail.
+// https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow
+func (c *ServiceCredentials) onBehalfOf(ctx context.Context, userToken string) (onBehalfOfToken string, expire time.Time, err error) {
+	params := make(url.Values)
+	params.Set("client_id", c.ClientID)
+	params.Set("client_secret", c.Secret)
+	params.Set("grant_type", "urn:ietf:params:oauth:grant-type:jwt-bearer")
+	params.Set("assertion", userToken)
+	params.Set("scope", c.ClusterURL+"/.default")
+	params.Set("requested_token_use", "on_behalf_of")
+	reqBody := strings.NewReader(params.Encode())
+
+	// https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-v2-protocols#endpoints
+	tokenURL := fmt.Sprintf("https://%s%s/oauth2/v2.0/token", authorityBaseURL(c.AzureCloud), url.PathEscape(c.TenantID))
```

Comment:
The `authorityBaseURL` func already resolves a url prefixed with `https://` so I believe you can remove the hard coded protocol from here.